### ISSUE TITLE
Remove useless interfaces

### ISSUE THREAD

**Issue by helmbold** [kotlintest/kotlintest#132]
Interfaces like `Eventually` which contains only a single (defined) method could be removed and the function could be made top-level and used directly instead.

```kotlin
interface Eventually {

  fun eventually(duration: Duration, f: () -> Unit): Unit {
    val end = System.nanoTime() + duration.nanoseconds
    var times = 0
    while (System.nanoTime() < end) {
      try {
        f()
        return
      } catch (e: Exception) {
        // ignore and proceed
      }
      times++
    }
    throw AssertionError("Test failed after ${duration.amount} ${duration.timeUnit}; attempted $times times")
  }
```

Shall we remove such interfaces?

**Comment by sksamuel** [comment]
I guess we can but it is a nice way of separating out the logic. If you remove the interface you just end up creating a bigger and bigger class. However, I would be in favour of making it automatically available to the end user.

**Comment by helmbold** [comment]
Are you aware of a way to make it available automatically? I thought of simple top-level functions. So instead of `MySpec : Eventually` you would import all functions from the Eventually.kt file via a star import. Since all top-level functions from different files in one package reside in the same name space, a single import for all matchers would be sufficient: `import io.kotlintest.matchers.*`.

This would also so solve the problem with reified generics in interface functions.

**Comment by helmbold** [comment]
I've just tried to remove these interfaces, but this leads to some errors. For example `should` can not be resolved in some cases. The best would be to rework the matchers first.

**Comment by sksamuel** [comment]
The way to make it automatically available is just to have the `Eventually` interface mixed in with the specs at the spec level. So the user extends a spec and gets Eventually automatically.

**Comment by helmbold** [comment]
I think it would be cleaner to make all matcher top-level functions and remove interfaces and inheritance entirely here. Having a matcher function at some unrelated class somewhere in our code base, to be able to use reified generics and to make it available in the scope, is not worth it (at least from my point of view).

**Comment by helmbold** [comment]
@sksamuel What do you think about making matchers top-level functions and letting the users import them explicitly? I think that would lead to a much better code structure. Or in the words of the Python community: [Explicit is better than implicit.](https://www.python.org/dev/peps/pep-0020/)

**Comment by sksamuel** [comment]
We can do. What's the benefit to the user.

**Comment by helmbold** [comment]
This suggestion is mainly about improving the internal code quality. However a benefit for the user could be more clearity, if he tries to understand the framework. And it wouldn't be obvious to have a certain matcher in a completely unrelated location just because of some language limitation (reified generics in this case). Another benefit could be the avoidance of name collisions when other matcher libs are used.

**Comment by aleixmorgadas** [comment]
@helmbold Once the #132 has been merged into release/2.0 branch, some errors related with the interfaces have appeared.

I was fixing it until I realized that I will finish undoing this work. Could you give it a look, please?
